File: src/main/java/day1java/exercise3/Console.java
package day1java.exercise3;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.function.Supplier;

public class Console<T> {

    public Console(Supplier<T> exec){
        this.exec = exec;
    }

    public Supplier<T> exec;


    public static Console<Void> printIO(String msg) {
        return new Console<>(() -> {
            System.out.println(msg);
            return null;
        });
    }

    public static Console<String> readlineIO() {
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        return new Console<String>(() -> {
            try {
                return reader.readLine();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });
    }

}

File: src/main/kotlin/day1/exercise3/Console.kt
package day1.exercise3

import java.io.BufferedReader
import java.io.InputStreamReader


data class Console<A> (val exec: () -> A) {

    fun andThen(other: Console<A>): Console<A> = Console {
        this.exec()
        other.exec()
    }

    fun <B> map(f: (A) -> B): Console<B> = Console {
        f(exec())
    }

    fun <B> flatMap(f: (A) -> Console<B>): Console<B> = Console {
        f(exec()).exec()
    }
}


fun printIO(msg: String): Console<Unit> = Console { println(msg)}

val reader =  BufferedReader( InputStreamReader(System.`in`))

fun readlineIO() = Console<String> { reader.readLine() }